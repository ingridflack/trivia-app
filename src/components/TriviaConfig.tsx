import {
  Button,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Grid,
  Select,
  useToast,
} from "@chakra-ui/react";
import { SubmitHandler, useForm } from "react-hook-form";
import { useEffect, useState } from "react";
import * as TriviaService from "../services/triviaService";
import * as UserService from "../services/userService";
import { Category, UserSearchSelectOption } from "../types/sharedTypes";
import { useNavigate } from "react-router-dom";
import {
  AMOUNT_OPTIONS,
  DIFFICULTY_OPTIONS,
  GAME_MODE_OPTIONS,
} from "../constants/trivia";
import AsyncSelect from "react-select/async";
import { MultiValue } from "react-select";

interface GameConfigValues {
  category: string;
  difficulty: string;
  amount: number;
  gameMode: string;
  type: string;
}

export default function TriviaConfig() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<GameConfigValues>({
    defaultValues: {
      category: "any",
      difficulty: "any",
      type: "any",
      amount: 5,
    },
  });

  const [categories, setCategories] = useState<Category[]>([]);
  const [gameMode, setGameMode] = useState("single");
  const navigate = useNavigate();
  const toast = useToast();
  const [invitedUsers, setInvitedUsers] = useState<
    MultiValue<UserSearchSelectOption>
  >([]);

  useEffect(() => {
    const fetchCategories = async () => {
      const { data } = await TriviaService.getCategories();
      setCategories(data.categories);
    };

    fetchCategories();
  }, []);

  const onSubmit: SubmitHandler<GameConfigValues> = async (values) => {
    try {
      const invitedUsersIds =
        gameMode === "multi"
          ? invitedUsers.map((option) => option.value)
          : undefined;

      const body = {
        ...values,
        invitedUsers: invitedUsersIds,
      };

      const { data } = await TriviaService.createTrivia(body);

      navigate(`/trivia/${data.triviaId}`);
    } catch {
      console.log("error");

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  const handleGameModeChange = (value: string) => {
    setGameMode(value);
  };

  const handleInviteUser = (value: MultiValue<UserSearchSelectOption>) => {
    setInvitedUsers(value);
  };

  const handleLoadUsers = async (username: string) => {
    const userData = localStorage.getItem("userData");
    const currentUserId = userData ? JSON.parse(userData).id : "";

    const { data } = await UserService.search({ username });
    return data
      .filter((user) => user._id !== currentUserId)
      .map((user) => ({ value: user._id, label: user.username }));
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmit)}
      style={{
        width: "100%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <Grid templateColumns="repeat(2, 1fr)" gap="0 20px">
        <FormControl isInvalid={!!errors.category} minW="186px">
          <FormLabel htmlFor="category" fontSize="small" marginBottom="4px">
            Category
          </FormLabel>
          <Select
            id="category"
            {...register("category", {
              required: "Category is required.",
            })}
          >
            <option value="any">Any category</option>
            {categories.map((category) => (
              <option key={category.id} value={category.id}>
                {category.name}
              </option>
            ))}
          </Select>
          <FormErrorMessage>
            {errors.category && errors.category.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.difficulty} minW="186px">
          <FormLabel htmlFor="difficulty" fontSize="small" marginBottom="4px">
            Difficulty
          </FormLabel>

          <Select
            id="difficulty"
            {...register("difficulty", {
              required: "Difficulty is required.",
            })}
          >
            {DIFFICULTY_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.difficulty && errors.difficulty.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.amount} minW="186px">
          <FormLabel
            marginTop="16px"
            htmlFor="amount"
            fontSize="small"
            marginBottom="4px"
          >
            Number of questions
          </FormLabel>

          <Select
            id="amount"
            {...register("amount", {
              required: "Amount is required.",
            })}
          >
            {AMOUNT_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.value}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.amount && errors.amount.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.gameMode} minW="186px">
          <FormLabel
            marginTop="16px"
            htmlFor="gameMode"
            fontSize="small"
            marginBottom="4px"
          >
            Game mode
          </FormLabel>

          <Select
            id="gameMode"
            {...register("gameMode", {
              required: "Game mode is required.",
            })}
            onChange={(e) => handleGameModeChange(e.target.value)}
          >
            {GAME_MODE_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.gameMode && errors.gameMode.message}
          </FormErrorMessage>
        </FormControl>
      </Grid>

      {gameMode === "single" ? (
        <Button
          colorScheme="purple"
          size="lg"
          width="100%"
          marginTop="16px"
          isLoading={isSubmitting}
          type="submit"
        >
          Start trivia
        </Button>
      ) : (
        <>
          <Flex mt={2} width="100%">
            <FormControl>
              <AsyncSelect
                isMulti
                cacheOptions
                loadOptions={handleLoadUsers}
                placeholder="Select user"
                options={[
                  { value: "user1", label: "User 1" },
                  { value: "user2", label: "User 2" },
                  { value: "user3", label: "User 3" },
                ]}
                onChange={handleInviteUser}
              />
            </FormControl>
          </Flex>

          {console.log({ invitedUsers, gameMode })}

          <Button
            colorScheme="purple"
            size="lg"
            width="100%"
            marginTop="16px"
            isLoading={isSubmitting}
            type="submit"
            isDisabled={invitedUsers.length === 0 && gameMode === "multi"}
          >
            Start game
          </Button>
        </>
      )}
    </form>
  );
}
