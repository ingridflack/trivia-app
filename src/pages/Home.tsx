import { Box, Button, Container, Image, Text } from "@chakra-ui/react";
import Header from "../components/Header";
import { Link } from "react-router-dom";

export default function Home() {
  return (
    <>
      <Header />
      <Box width="100%" backgroundColor="#AE79F8" overflow="hidden">
        <Container paddingY="80px" position="relative">
          <Text
            fontSize="6xl"
            lineHeight="1.2"
            fontWeight="bold"
            color="white"
            maxWidth="550"
          >
            Challenge yourself in a trivia match!
          </Text>

          <Button
            backgroundColor="green.400"
            color="white"
            width="180px"
            _hover="green.400"
            marginTop="38px"
            as={Link}
            to="/trivia/config"
          >
            Start a trivia
          </Button>
          <Image
            src="/src/assets/home.svg"
            alt="Home image"
            position="absolute"
            right="-160px"
            bottom="-60px"
            width="674px"
            height="418px"
            pointerEvents="none"
          />
        </Container>
      </Box>
    </>
  );
}
