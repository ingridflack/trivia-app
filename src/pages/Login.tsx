import {
  Box,
  Card,
  Container,
  Image,
  Link,
  Text,
  useToast,
} from "@chakra-ui/react";
import LoginForm from "../components/LoginForm";
import Header from "../components/Header";
import { useNavigate } from "react-router-dom";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../services/authService";
import useAuth from "../hooks/useAuth";

export interface LoginFormValues {
  email: string;
  password: string;
}

export default function Login() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormValues>();
  const navigate = useNavigate();
  const toast = useToast();
  const { saveUserData } = useAuth();

  const onSubmit: SubmitHandler<LoginFormValues> = async (values) => {
    try {
      const { data } = await AuthService.login(values);

      saveUserData(data);

      navigate("/");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding="70px"
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Sign in
            </Text>
            <LoginForm
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              isSubmitting={isSubmitting}
              register={register}
            />

            <Box marginTop="20px" display="flex" justifyContent="space-between">
              <Link
                fontSize="small"
                color="gray.500"
                href="/trivia/recovery-email"
              >
                Forgot your password?
              </Link>

              <Link fontSize="small" color="gray.500" href="/sign-up">
                Create an account
              </Link>
            </Box>
          </Box>

          <Image
            src="src/assets/login.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
          />
        </Card>
      </Container>
    </>
  );
}
