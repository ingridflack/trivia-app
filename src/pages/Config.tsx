import { Box, Card, Container, Image, Text } from "@chakra-ui/react";
import TriviaConfig from "../components/TriviaConfig";
import Header from "../components/Header";
import { Footer } from "../components/Footer";

export default function Config() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/src/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
