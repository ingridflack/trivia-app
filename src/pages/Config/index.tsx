import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import TriviaConfig from "../../components/TriviaConfig";
import Header from "../../components/Header";
import { Footer } from "../../components/Footer";
import * as UserService from "../../services/userService";
import * as TriviaService from "../../services/triviaService";
import { SubmitHandler, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useState } from "react";
import { MultiValue } from "react-select";
import { UserSearchSelectOption } from "../../types/sharedTypes";
import useAuth from "../../hooks/useAuth";

export interface GameConfigValues {
  category: string;
  difficulty: string;
  amount: number;
  gameMode: string;
  type: string;
}

export default function Config() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<GameConfigValues>({
    defaultValues: {
      difficulty: "any",
      type: "any",
      amount: 5,
    },
  });

  const [gameMode, setGameMode] = useState("single");
  const navigate = useNavigate();
  const toast = useToast();
  const [invitedUsers, setInvitedUsers] = useState<
    MultiValue<UserSearchSelectOption>
  >([]);
  const { userData } = useAuth();

  const onSubmit: SubmitHandler<GameConfigValues> = async (values) => {
    try {
      const invitedUsersIds =
        gameMode === "multi"
          ? invitedUsers.map((option) => option.value)
          : undefined;

      const body = {
        ...values,
        invitedUsers: invitedUsersIds,
      };

      const { data } = await TriviaService.createTrivia(body);

      navigate(`/trivia/${data.triviaId}`);
    } catch {
      console.log("error");

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  const handleGameModeChange = (value: string) => {
    setGameMode(value);
  };

  const handleInviteUser = (value: MultiValue<UserSearchSelectOption>) => {
    setInvitedUsers(value);
  };

  const handleLoadUsers = async (username: string) => {
    const currentUserId = userData ? userData.id : "";

    const { data } = await UserService.search({ username });
    return data
      .filter((user) => user._id !== currentUserId)
      .map((user) => ({ value: user._id, label: user.username }));
  };

  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig
              onLoadUsers={handleLoadUsers}
              onInviteUser={handleInviteUser}
              onGameModeChange={handleGameModeChange}
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              register={register}
              isSubmitting={isSubmitting}
              gameMode={gameMode}
              invitedUsers={invitedUsers}
            />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
