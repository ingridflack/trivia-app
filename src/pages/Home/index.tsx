import {
  Box,
  Button,
  Container,
  Heading,
  Image,
  Skeleton,
  Stack,
  Text,
} from "@chakra-ui/react";
import Header from "../../components/Header";
import { Link, useNavigate } from "react-router-dom";
import { PendingTriviaList } from "../../components/PendingTriviaList";
import { useEffect, useState } from "react";
import * as TriviaService from "../../services/triviaService";
import { CreateTriviaBody, PendingTrivia } from "../../types/sharedTypes";
import { Footer } from "../../components/Footer";
import useAuth from "../../hooks/useAuth";
import { TRIVIA_CATEGORIES } from "../../constants/trivia";
import { CategoryCard } from "../../components/CategoryCard";

export default function Home() {
  const [pendingTrivia, setPendingTrivia] = useState<PendingTrivia[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const { userData } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    const fetchPendingTrivia = async () => {
      try {
        const { data } = await TriviaService.getPendingTrivia();
        setPendingTrivia(data.pendingTrivia);
      } catch (error) {
        console.log(error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchPendingTrivia();
  }, []);

  const handleCategoryCardClick = async (params: CreateTriviaBody) => {
    try {
      if (!userData) {
        localStorage.setItem("triviaConfig", JSON.stringify(params));
        navigate("/login");
        return;
      }

      const { data } = await TriviaService.createTrivia(params);

      navigate(`/trivia/${data.triviaId}`);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Header />

      <Box as="main">
        <Box width="100%" backgroundColor="purple.500" overflow="hidden">
          <Container
            padding={{ base: "40px 20px", md: "80px 0" }}
            position="relative"
          >
            <Text
              fontSize={{ base: "3xl", md: "6xl" }}
              lineHeight="1.2"
              fontWeight="bold"
              color="white"
              maxWidth="550"
            >
              Challenge yourself in a trivia match!
            </Text>

            <Button
              backgroundColor="green.400"
              color="white"
              width="180px"
              _hover="green.400"
              marginTop="38px"
              as={Link}
              to="/trivia/config"
            >
              Start a trivia
            </Button>
            <Image
              src="/assets/home.svg"
              alt="Home image"
              position="absolute"
              right="-160px"
              bottom="-60px"
              width="674px"
              height="418px"
              pointerEvents="none"
              display={{ base: "none", md: "none", lg: "block" }}
            />
          </Container>
        </Box>

        {userData && (
          <Container padding={{ base: "40px 20px", md: "80px 0 120px" }}>
            <Box as="header" marginBottom={{ base: "20px", md: "40px" }}>
              <Heading fontSize="3xl">Pending trivias</Heading>
            </Box>

            {isLoading ? (
              <Stack
                display="grid"
                gap="16px"
                gridTemplateColumns={{
                  base: "repeat(1fr)",
                  md: "repeat(4, 1fr)",
                }}
              >
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
              </Stack>
            ) : pendingTrivia.length ? (
              <PendingTriviaList pendingTriviaList={pendingTrivia} />
            ) : (
              <Text>No pending trivias</Text>
            )}
          </Container>
        )}

        <Box
          as="section"
          padding={{ base: "40px 20px", md: "80px 0 120px" }}
          backgroundColor="gray.100"
          boxShadow="0px 30px 100px 0 rgba(39, 39, 39, 0.2);"
        >
          <Box
            as="header"
            textAlign="center"
            marginBottom={{ base: "20px", md: "60px" }}
          >
            <Heading fontSize={{ base: "3xl", md: "5xl" }} marginBottom="20px">
              Show off your knowledge
            </Heading>
            <Text fontSize={{ base: "md", md: "xl" }}>
              Take a quiz and impress your friends with what you know!
            </Text>
          </Box>

          <Container
            display="grid"
            gridTemplateColumns={{ base: "repeat(1fr)", md: "repeat(5, 1fr)" }}
            gap="16px"
            padding={{ base: "0", md: "0 0" }}
          >
            {Object.values(TRIVIA_CATEGORIES).map((category) => (
              <CategoryCard
                key={category.title}
                title={category.title}
                icon={category.icon}
                color={category.color}
                onCardClick={() => handleCategoryCardClick(category.params)}
              />
            ))}
          </Container>
        </Box>
      </Box>

      <Footer />
    </>
  );
}
