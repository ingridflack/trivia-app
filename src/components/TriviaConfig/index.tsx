import {
  Button,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Grid,
  Select,
} from "@chakra-ui/react";
import { FieldErrors, UseFormRegister } from "react-hook-form";

import { UserSearchSelectOption } from "../../types/sharedTypes";
import {
  AMOUNT_OPTIONS,
  DIFFICULTY_OPTIONS,
  GAME_MODE_OPTIONS,
  TRIVIA_CATEGORIES,
} from "../../constants/trivia";
import AsyncSelect from "react-select/async";
import { MultiValue } from "react-select";
import { GameConfigValues } from "../../pages/Config";

export interface TriviaConfigProps {
  onLoadUsers: (inputValue: string) => Promise<UserSearchSelectOption[]>;
  onInviteUser: (value: MultiValue<UserSearchSelectOption>) => void;
  onGameModeChange: (gameMode: string) => void;
  onSubmit: (e?: React.BaseSyntheticEvent) => Promise<void>;
  errors: FieldErrors<GameConfigValues>;
  register: UseFormRegister<GameConfigValues>;
  isSubmitting: boolean;
  gameMode: string;
  invitedUsers: MultiValue<UserSearchSelectOption>;
}

export default function TriviaConfig({
  onLoadUsers,
  onInviteUser,
  onGameModeChange,
  onSubmit,
  errors,
  register,
  isSubmitting,
  gameMode,
  invitedUsers,
}: TriviaConfigProps) {
  return (
    <form
      role="form"
      onSubmit={onSubmit}
      style={{
        width: "100%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <Grid
        templateColumns={{ base: "1fr", md: "repeat(2, 1fr)" }}
        gap="0 20px"
        width="100%"
      >
        <FormControl isInvalid={!!errors.category} minW="186px">
          <FormLabel htmlFor="category" fontSize="small" marginBottom="4px">
            Category
          </FormLabel>
          <Select
            id="category"
            {...register("category", {
              required: "Category is required.",
            })}
          >
            {Object.entries(TRIVIA_CATEGORIES).map(([id, { title }]) => (
              <option key={id} value={id}>
                {title}
              </option>
            ))}
          </Select>
          <FormErrorMessage>
            {errors.category && errors.category.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.difficulty} minW="186px">
          <FormLabel
            htmlFor="difficulty"
            marginTop={{ base: "16px", md: 0 }}
            fontSize="small"
            marginBottom="4px"
          >
            Difficulty
          </FormLabel>

          <Select
            id="difficulty"
            {...register("difficulty", {
              required: "Difficulty is required.",
            })}
          >
            {DIFFICULTY_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.difficulty && errors.difficulty.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.amount} minW="186px">
          <FormLabel
            marginTop="16px"
            htmlFor="amount"
            fontSize="small"
            marginBottom="4px"
          >
            Number of questions
          </FormLabel>

          <Select
            id="amount"
            {...register("amount", {
              required: "Amount is required.",
            })}
          >
            {AMOUNT_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.value}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.amount && errors.amount.message}
          </FormErrorMessage>
        </FormControl>

        <FormControl isInvalid={!!errors.gameMode} minW="186px">
          <FormLabel
            marginTop="16px"
            htmlFor="gameMode"
            fontSize="small"
            marginBottom="4px"
          >
            Game mode
          </FormLabel>

          <Select
            id="gameMode"
            {...register("gameMode", {
              required: "Game mode is required.",
            })}
            onChange={(e) => onGameModeChange(e.target.value)}
          >
            {GAME_MODE_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </Select>

          <FormErrorMessage>
            {errors.gameMode && errors.gameMode.message}
          </FormErrorMessage>
        </FormControl>
      </Grid>

      {gameMode === "single" ? (
        <Button
          colorScheme="purple"
          size="lg"
          width="100%"
          marginTop="16px"
          isLoading={isSubmitting}
          type="submit"
        >
          Start trivia
        </Button>
      ) : (
        <>
          <Flex mt={2} width="100%">
            <FormControl>
              <AsyncSelect
                isMulti
                cacheOptions
                loadOptions={onLoadUsers}
                placeholder="Select user"
                options={[
                  { value: "user1", label: "User 1" },
                  { value: "user2", label: "User 2" },
                  { value: "user3", label: "User 3" },
                ]}
                onChange={onInviteUser}
              />
            </FormControl>
          </Flex>

          <Button
            colorScheme="purple"
            size="lg"
            width="100%"
            marginTop="16px"
            isLoading={isSubmitting}
            type="submit"
            isDisabled={invitedUsers.length === 0 && gameMode === "multi"}
          >
            Start game
          </Button>
        </>
      )}
    </form>
  );
}
