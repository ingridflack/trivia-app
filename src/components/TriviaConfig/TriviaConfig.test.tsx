import { cleanup, fireEvent, render } from "@testing-library/react";
import { afterEach, describe, expect, it, vi } from "vitest";
import TriviaConfig, { TriviaConfigProps } from ".";

import { FieldErrors } from "react-hook-form";
import { GameConfigValues } from "../../pages/Config";

const formFieldsData = [
  {
    errors: { category: { message: "Invalid category" } },
    name: "category",
    regex: /invalid category/i,
  },
  {
    errors: { amount: { message: "amount is required" } },
    name: "amount",
    regex: /amount is required/i,
  },
  {
    errors: { gameMode: { message: "Invalid gameMode" } },
    name: "gameMode",
    regex: /invalid gameMode/i,
  },
  {
    errors: { difficulty: { message: "Invalid difficulty" } },
    name: "difficulty",
    regex: /invalid difficulty/i,
  },
];

const setup = (props: Partial<TriviaConfigProps> = {}) =>
  render(
    <TriviaConfig
      onLoadUsers={vi.fn()}
      onInviteUser={vi.fn()}
      onGameModeChange={vi.fn()}
      onSubmit={vi.fn()}
      errors={{}}
      isSubmitting={false}
      register={vi.fn()}
      gameMode="single"
      invitedUsers={[]}
      {...props}
    />
  );

describe("TriviaConfig", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match the snapshot", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("renders category select", () => {
    const { getByLabelText } = setup();
    expect(getByLabelText(/Category/i)).toBeTruthy();
  });

  it("renders difficulty select", () => {
    const { getByLabelText } = setup();
    expect(getByLabelText(/Difficulty/i)).toBeTruthy();
  });

  it("renders amount select", () => {
    const { getByLabelText } = setup();
    expect(getByLabelText(/Number of questions/i)).toBeTruthy();
  });

  it("renders game mode select", () => {
    const { getByLabelText } = setup();
    expect(getByLabelText(/Game mode/i)).toBeTruthy();
  });

  it("renders Start trivia button in single mode", () => {
    const { getByRole } = setup();
    expect(getByRole("button", { name: /Start trivia/i })).toBeTruthy();
  });

  it("renders Start trivia button in multi mode", () => {
    const { getByRole } = setup({ gameMode: "multi" });
    expect(getByRole("button", { name: /Start game/i })).toBeTruthy();
  });

  it("should call onSubmit when the form is submitted", () => {
    const mockOnSubmit = vi.fn();
    const { getByRole } = setup({ onSubmit: mockOnSubmit });

    fireEvent.submit(getByRole("form"));
    expect(mockOnSubmit).toHaveBeenCalled();
  });

  it("calls onGameModeChange when game mode is changed", () => {
    const onGameModeChange = vi.fn();
    const { getByRole, getByLabelText } = setup({ onGameModeChange });

    const gameModeSelect = getByLabelText(/Game mode/i, { selector: "select" });
    fireEvent.change(gameModeSelect, { target: { value: "multi" } });

    const startTriviaButton = getByRole("button", { name: /Start trivia/i });
    fireEvent.click(startTriviaButton);

    expect(onGameModeChange).toHaveBeenCalledWith("multi");
  });

  formFieldsData.forEach(({ errors, name, regex }) => {
    it(`should display the ${name} error message when it's invalid`, () => {
      const props = { errors: errors as FieldErrors<GameConfigValues> };
      const { getByText } = setup(props);
      expect(getByText(regex)).toBeTruthy();
    });
  });
});
