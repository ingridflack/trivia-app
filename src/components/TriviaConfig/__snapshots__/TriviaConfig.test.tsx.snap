// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TriviaConfig > should match the snapshot 1`] = `
<div>
  <form
    role="form"
    style="width: 100%; display: flex; align-items: center; justify-content: center; flex-direction: column;"
  >
    <div
      class="css-1ggap6e"
    >
      <div
        class="chakra-form-control css-1u79ci6"
        role="group"
      >
        <label
          class="chakra-form__label css-t3kal5"
          for="category"
          id="field-:r0:-label"
        >
          Category
        </label>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-1ik61og"
            id="category"
          >
            <option
              value="9"
            >
              General Knowledge
            </option>
            <option
              value="10"
            >
              Books
            </option>
            <option
              value="14"
            >
              Television
            </option>
            <option
              value="19"
            >
              Mathematics
            </option>
            <option
              value="31"
            >
              Japanese Anime & Manga
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="chakra-form-control css-1u79ci6"
        role="group"
      >
        <label
          class="chakra-form__label css-t3kal5"
          for="difficulty"
          id="field-:r1:-label"
        >
          Difficulty
        </label>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-1ik61og"
            id="difficulty"
          >
            <option
              value="any"
            >
              Any
            </option>
            <option
              value="easy"
            >
              Easy
            </option>
            <option
              value="medium"
            >
              Medium
            </option>
            <option
              value="hard"
            >
              Hard
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="chakra-form-control css-1u79ci6"
        role="group"
      >
        <label
          class="chakra-form__label css-19h9qxr"
          for="amount"
          id="field-:r2:-label"
        >
          Number of questions
        </label>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-1ik61og"
            id="amount"
          >
            <option
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="15"
            >
              15
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="chakra-form-control css-1u79ci6"
        role="group"
      >
        <label
          class="chakra-form__label css-19h9qxr"
          for="gameMode"
          id="field-:r3:-label"
        >
          Game mode
        </label>
        <div
          class="chakra-select__wrapper css-42b2qy"
        >
          <select
            class="chakra-select css-1ik61og"
            id="gameMode"
          >
            <option
              value="single"
            >
              Single player
            </option>
            <option
              value="multi"
            >
              Multiplayer
            </option>
          </select>
          <div
            class="chakra-select__icon-wrapper css-162mkon"
          >
            <svg
              aria-hidden="true"
              class="chakra-select__icon"
              focusable="false"
              role="presentation"
              style="width: 1em; height: 1em; color: currentColor;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <button
      class="chakra-button css-1w54gy2"
      type="submit"
    >
      Start trivia
    </button>
  </form>
</div>
`;
